### PR TITLE
The initial tools and conventions for the new documentation.

### PR DESCRIPTION
**Pull request by HU90m**
Sorry that there is so much content in one PR. The primary reason for this is to ensure there is no regression in the documentation during the transition to a new system. This PR has the minimal number of changes such that the new documentation site has all the content of the current documentation.

No content has therefore been changed significantly. Content and layout improvements have been explored by other contributors, but will be submitted in separate PRs. This PR also contains no styling or landing page improvements, such the interactive block diagram that has been created which, again, will be submitted in a different PR.

The only changes to content are to update files to adhere to the proposed new conventions. These conventions mainly affect the position of content. Care has been taken to ensure these moves were all picked up by `git`; the commits where these moves occur have a `[doc mv]` plug.

The main changes are the addition of [`mdbook`](https://rust-lang.github.io/mdBook/) (the proposed documentation generator) and it's accompanying preprocessors. These tools set us up nicely for the content changes that have been discussed previously. This is how it looks:

Screenshot from 2023-02-20 14-58-59

To try it out run:

```sh
./util/site/build-docs.sh serve
```

### File Structure

When contributing to the current documentation there was confusion about where certain documentation should live. For example, `reggen` tool had it's documentation in both [the reference manual](https://docs.opentitan.org/doc/rm/register_tool/) and [the utilities section](https://docs.opentitan.org/util/reggen/doc/). A solution to this is always having documentation next to what it is documenting.

The proposed convention is to ask that the documentation for a particular thing live in the `README.md` of that things directory. If there is additional documentation related to this thing that wouldn't make sense in any of it's subdirectorys, this additional documentation would live within a general `doc` sub-directory. This extends to the root of the directory where `doc` contains documentation (and guides) that cover the project as a whole.

The use of `README.md` reduces duplication that exists between `_index.md`s and `README.md`s in the current repo (for example `sw/device/lib/dif/{_index.md,README.md}`). Where there isn't duplication, it is likely that the content in `README.md` would be useful in the documentation.

Practically speaking, this convention is what was already widely done in the repository (with the exception of using `README.md`s), such as the `doc` directory. Most of the movement of files where to make all files follow this convention.

The only other thing to note is that the current docs prefers snake case for documentation files. This convention has been adhered to.

### File Content

The main changes to content have happened as a result of moving away from [hugo](https://gohugo.io/). As we have done this, we have tried to move as much as possible to standard [commonmark](https://commonmark.org/) and, furthermore, to a format that is conducive to reading the content in github and in plain-text.

Examples of these changes are below:

<table>
<tr><td> Before </td> <td> After </td></tr>
<tr><td>

  ````markdown
[click me]({{< relref ".." >}})
  ````

</td><td>

  ````markdown
[click me](../README.md)
  ````

</td></tr>
<tr><td>

  ````markdown
  {{< wavejson >}}
  _content_
  {{</wavejson>}}
  ````

</td><td>

  ````markdown
  ```wavejson
  _content_
  ```
  ````

</td></tr>
<tr><td>

  ````markdown
{{< regref "adc_pd_ctl" >}}
  ````

</td><td>

  ````markdown
[`ADC_PD_CTL`](data/adc_ctrl.hjson#adc_pd_ctl)
  ````

</td></tr>
</table>

### Documentation Layout

There was a clear feedback from people using the documentation that it needs to be better targeted to it's audience. As stated, this PR doesn't introduce new content. However, it does set the documentation generation up for new content so, as the files where moved into a consistent structure, where it made sense, files where organised with respect to their target audience. An example of this is the new 'contributing' section. 

The `doc/guides` directory has also been created to hold how-to guides or tutorials targeting specific user groups and which don't make sense in the main book.

This layout will almost definitely evolve as new content is added.

### Books

An `mdbook` takes it's settings from a `book.toml` in it's root directory and renders all files present in `SUMMARY.md` (other files are copied across without being rendered). There are currently two books: the main book in `.` and the getting started guide in `doc/guides/getting_started`.

### Preprocessors

Where needed, old hugo shortcodes have been replaced with preprocessors. These preprocessors live in `./utils/mdbook_*.py` and are configured from within the `book.toml`. They use the standard (for mdbook) `{{#foo bar}}` template in the markdown where standard commonmark can't be used. Many preprocessors can likely be phased out as content changes or by leaning into `AUTOGEN` based approaches (explained next).

### AUTOGEN

There are may parts of the documentation that make sense to and thus are generated automatically, for example register tables. This is because the content is specified somewhere else in the repository and by generating content from this source one doesn't have to worry about differences/inaccuracies.

Automatically generated content has the added bonus over a templating approach (e.g. `{{< incGenFromIpDesc ... >}}`) of being useful when read in github's rendering or in plain-text. The generation and checking of this content needs to be automated though to ensure it doesn't fall out of sync with the project.

A starting point has been used for the testplans (see `hw/top_earlgrey/doc/design/README.md`) and 'selfdoc'. The generated content is delimited by `<!-- BEGIN AUTOGEN cmd -->` and `<!-- END AUTOGEN -->` lines, `cmd` being the command used to generate the command. The script `utils/autogen.py` looks through markdown files, running the commands specified and updating content with their standard output. This approach was decided upon because it shows clearly the provenance of the content and it is simple to understand without lacking flexibility.

This script will definitely need to be set up in CI to test that content is re-generated as it's source information is changed. There are also plans to add a context aware linter that can find pages that should have certain content but don't.

## Closing Remarks

Many people have worked on very exiting additions to the documentation that are not part of this PR, but that will slowly trickle their way into the repository in a more digestible manner. There is also a mock-up of the proposed new documentation site with content changes, a modified landing page, a fancy new block diagram, and more that we hope to host soon so that people can see the vision for the new documentation and provide feedback.

Edit: [The demo site is live!](https://staging.opentitan.org/)

### CODE REVIEW

File: util/site/build-docs.sh [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+set -e
+
+command="build"
+
+case "$1" in
+  "build"|"build-local"|"serve"|"serve-proxy")
+    command="$1"
+    public_url="${2}"
+    public_port="${3}"
```
Comment (jwnrt) [inline util/site/build-docs.sh:11]: These args don't have documentation or validation

File: util/site/build-docs.sh [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+getURLs
+
+# Export some environment variables that tools will pick up
+export HUGO_PARAMS_DOCSURL="${docs_url}" # hugo
+export URL_ROOT="${docs_url}" # earlgrey_diagram
+
+# Build up doxygen command
+doxygen_env="env"
+doxygen_env+=" SRCTREE_TOP=${proj_root}"
+doxygen_env+=" DOXYGEN_OUT=${build_dir}/gen"
```
Comment (jwnrt) [inline util/site/build-docs.sh:92]: Mixture of `export` and `env` - I think the `export`s are easier to read personally.

File: util/site/build-docs.sh [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+# Each book should only be passed the preprocessors it specifies inside the book.toml
+# ./book.toml
+book_env="env"
+book_env+=" MDBOOK_PREPROCESSOR__TESTPLAN__COMMAND=${proj_root}/util/mdbook_testplan.py"
+book_env+=" MDBOOK_PREPROCESSOR__OTBN__COMMAND=${proj_root}/util/mdbook_otbn.py"
+book_env+=" MDBOOK_PREPROCESSOR__DOXYGEN__COMMAND=${proj_root}/util/mdbook_doxygen.py"
+book_env+=" MDBOOK_PREPROCESSOR__REGGEN__COMMAND=${proj_root}/util/mdbook_reggen.py"
+book_env+=" MDBOOK_PREPROCESSOR__WAVEJSON__COMMAND=${proj_root}/util/mdbook_wavejson.py"
+book_env+=" MDBOOK_PREPROCESSOR__README2INDEX__COMMAND=${proj_root}/util/mdbook_readme2index.py"
+book_env+=" MDBOOK_PREPROCESSOR__DASHBOARD__COMMAND=${proj_root}/util/mdbook_dashboard.py"
```
Comment (jwnrt) [inline util/site/build-docs.sh:115]: Aren't these already set in `book.toml`?

**Review by jwnrt — COMMENTED** [review]
Nice work! Not a full review (this is a big change), just some comments on the build.

Also I think `book.toml` and `SUMMARY.md` should probably go into `doc/`. The repo root is already quite cluttered.

**Review by andreaskurth — COMMENTED** [review]
The changes overall LGTM, thanks for all this work to everyone involved!  I think it's very valuable (1) to have Markdown files that are more readable directly in the repository (on GitHub or in the code editor) and (2) to have documentation located close to the code it is documenting and then target different audiences as part of rendering the documentation.

With >10k lines changed, I could obviously not review every single line, but while scrolling through the changes I did not spot obvious mistakes.  Thanks for the effort of keeping the diff clean even though many lines have changed.

The lint check failures (see CI logs) will need to be fixed.

Regarding commit splitting, I wonder how atomic the individual commits are?  That is, is the repository in a working state after every commit in this PR?  If not, we may prefer squashing commits for which this does not hold.

Git still shows some files as deleted rather than moved.  One example is `doc/rm/crossbar_tool/index.md`.  If I understand correctly, the content of that file has been moved elsewhere.  I don't think there's a problem with that, but with so many changes how can we be sure that no content was lost in this reorganization?

File: doc/contributing/hw/methodology.md [hunk 4]
```diff
@@ -226,7 +226,7 @@ These are discussed in the [Getting Started Designing Hardware](./design.md) doc
 ## FPGA vs Silicon
 
 One output of the OpenTitan project will be silicon instantiations of hardware functionality described in this open source repository.
-The RTL repository defines design functionality at a level satisfactory to prove the hardware and software functionality in an FPGA (see [user guides]({{< relref "doc/ug" >}})).
+The RTL repository defines design functionality at a level satisfactory to prove the hardware and software functionality in an FPGA (see [user guides](https://docs.opentitan.org/doc/guides/getting_started/)).
```
Comment (marnovandermaas) [inline doc/contributing/hw/methodology.md:227]: user guides -> getting started guide

File: doc/introduction.md [hunk 5]
```diff
@@ -35,7 +35,7 @@ For questions about how the project is organized, see the [project](./project_go
 ## Development
 
 * [Getting Started](./guides/getting_started/src/README.md)
-* [User Guides]({{< relref "doc/ug" >}})
+* [User Guides](https://docs.opentitan.org/doc/guides/getting_started/)
```
Comment (marnovandermaas) [inline doc/introduction.md:36]: These don't exist anymore in the new world. Most of this has moved to contributing.

File: doc/introduction.md [hunk 6]
```diff
@@ -35,7 +35,7 @@ For questions about how the project is organized, see the [project](./project_go
 ## Development
 
 * [Getting Started](./guides/getting_started/src/README.md)
```
Comment (marnovandermaas) [inline doc/introduction.md:35]: This should also be a docs.opentitan.org link right?

File: util/check_links.py [hunk 7]
```diff
@@ -1,653 +0,0 @@
-#!/usr/bin/env python3
```
Comment (marnovandermaas) [inline util/check_links.py]: Can you just remove this check_links.py from the previous commit?

File: doc/project_governance/development_stages.md [hunk 8]
```diff
@@ -1,7 +1,4 @@
----
-aliases: [/doc/project/hw_stages/]
```
Comment (marnovandermaas) [inline doc/project_governance/development_stages.md:2]: Do we not need these aliases anymore?

File: hw/dv/doc/dv_doc_template.md [hunk 9]
```diff
@@ -1,7 +1,4 @@
----
-title: "FOO DV document"
-draft: true
```
Comment (marnovandermaas) [inline hw/dv/doc/dv_doc_template.md:3]: This is a draft file. Maybe add a comment like:
```html
<!-- This is a draft MD file and should not be rendered directly on the webiste. -->
```

File: hw/formal/README.md [hunk 10]
```diff
@@ -1,7 +1,4 @@
----
-title: "OpenTitan Assertions"
-unlisted: true
```
Comment (marnovandermaas) [inline hw/formal/README.md:3]: This file is unlisted. How are we guaranteeing that? Maybe we should add a comment in the file?

File: hw/formal/README.md [hunk 11]
```diff
@@ -1,7 +1,4 @@
...
----
+# OpenTitan Assertions
 
 # OpenTitan Assertions
```
Comment (marnovandermaas) [inline hw/formal/README.md:3]: This section heading has the same importance as the title.

File: hw/ip/adc_ctrl/README.md [hunk 12]
```diff
@@ -1,6 +1,4 @@
----
-title: "Analog to Digital Converter Control Interface"
----
+# Analog to Digital Converter Control Interface
 
 # Overview
```
Comment (marnovandermaas) [inline hw/ip/adc_ctrl/README.md:3]: I had made a commit somewhere that removes all the `# Overview`
This stops many of the section heading of having the same importance as the title.

File: util/mdbook-doxygen [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            "\n<details><summary>\nGenerated from <a href=\"{}\">{}</a></summary>\n".format(
+                site_url / src_path,
+                file_name,
+            ),
+        )
+        buffer.write("\n```c\n{}\n```\n".format(chapter["content"]))
+        buffer.write("</details>")
+        chapter["content"] = buffer.getvalue()
+
+        chapter["path"] = chapter["path"].removesuffix(".h") + "_h.html"
```
Comment (marnovandermaas) [inline util/mdbook-doxygen]: The fix that you have later for this should be squashed into here: https://github.com/lowRISC/opentitan/pull/17323/commits/4d2b14a5883a0894d405445766565244e726e212

File: util/mdbook-reggen [hunk 14]
```diff
@@ -19,16 +19,11 @@ from reggen.ip_block import IpBlock
 import reggen.gen_cfg_html as gen_cfg_html
```
Comment (marnovandermaas) [inline util/mdbook-reggen]: Is there any reason why this commit cannot be squashed into the previous reggen commit?
https://github.com/lowRISC/opentitan/pull/17323/commits/0a1340921d27990c5dc8291434b3ccc307db044f

File: SUMMARY.md [hunk 15]
```diff
@@ -14,13 +14,16 @@
   - [Alert Handler](./hw/top_earlgrey/ip_autogen/alert_handler/README.md)
     - [Design Verification](./hw/top_earlgrey/ip_autogen/alert_handler/dv/README.md)
     - [Interface and Registers](./hw/top_earlgrey/ip_autogen/alert_handler/data/alert_handler.hjson)
+    - [Device Interface Functions](./sw/device/lib/dif/dif_alert_handler.h)
```
Comment (marnovandermaas) [inline SUMMARY.md:18]: For this commit message and others that start with a verb, I think the convention is to capitalize the first letter after the tag.

File: SUMMARY.md [hunk 16]
```diff
@@ -13,11 +13,13 @@
   - [Analog Sensor Top](./hw/top_earlgrey/ip/ast/README.md)
   - [Alert Handler](./hw/top_earlgrey/ip_autogen/alert_handler/README.md)
     - [Design Verification](./hw/top_earlgrey/ip_autogen/alert_handler/dv/README.md)
+      - [Testplan](./hw/top_earlgrey/ip_autogen/alert_handler/data/alert_handler_testplan.hjson)
```
Comment (marnovandermaas) [inline SUMMARY.md:16]: I'm not sure if I am a fan of having just the one subsection in most of these Design Verification sections. It's usually nice to have at least two subsections.

File: hw/top_earlgrey/dv/README.md [hunk 17]
```diff
@@ -139,8 +139,8 @@ For a list of available tests  to run, please see the 'Tests' column in the [tes
 ### Nightly
 
 ## Testplan (RTL simulations)
-{{< incGenFromIpDesc "../../data/chip_testplan.hjson" "testplan" >}}
+[Testplan](../data/chip_testplan.hjson)
 
 ## Testplan (Gate level simulations)
 Note that the descriptions of the test below may be replicated from the table above.
```
Comment (marnovandermaas) [inline hw/top_earlgrey/dv/README.md:145]: This note can be removed.

File: util/mdbook-reggen [hunk 18]
```diff
@@ -95,7 +95,7 @@ def main() -> None:
 
             name = reg[1] + "." + reg[2] if len(reg) == 3 else reg[1]
 
-            return "[`{}`]({}#Reg_{})".format(name, path, reg[1].lower())
```
Comment (marnovandermaas) [inline util/mdbook-reggen]: Is this something that can be squashed into the commit that introduces the reggen preprocessor?

File: util/mdbook-reggen [hunk 19]
```diff
@@ -75,30 +73,6 @@ def main() -> None:
...
-                )
-            except KeyError:
-                sys.exit(f"Ip block with name '{reg[0]}' could not be found.")
-
-            name = reg[1] + "." + reg[2] if len(reg) == 3 else reg[1]
-
-            return "[`{}`]({}#{})".format(name, path, reg[1].lower())
-
-        chapter["content"] = REGREF_PATTERN.sub(regref_swap, chapter["content"])
-
```
Comment (marnovandermaas) [inline util/mdbook-reggen]: Does this just remove the automatic generation that was needed previously to convert the old to the new docs?

File: book.toml [hunk 20]
```diff
@@ -22,29 +22,29 @@ additional-css = [
 ]
 
 [preprocessor.readme2index]
-command = "./util/mdbook-readme2index"
+command = "./util/mdbook_readme2index.py"
```
Comment (marnovandermaas) [inline book.toml:26]: Can this commit be squashed into the individual commits where these preprocessors are introduced?

File: book.toml [hunk 21]
```diff
@@ -6,7 +6,8 @@ src = "."
 title = "OpenTitan Documentation"
 
 [output.html]
-fold = { enable = true, level = 1 }
+site-url = "book/"
+fold = { enable = true}
```
Comment (marnovandermaas) [inline book.toml:10]: Can't these changes be added in the previous commit where book.toml was introduced?

**Review by marnovandermaas — COMMENTED** [review]
Thanks for this amazing PR and to everyone who contributed to it! I had a few comments where I saw some issues or had some suggestions.

In general I like the idea of having separate move commits, even if they temporarily disrupt the documentation setup. It is unlikely that someone will want to run git bisect on documentation right?

**Comment by HU90m** [comment]
> Regarding commit splitting, I wonder how atomic the individual commits are? That is, is the repository in a working state after every commit in this PR? If not, we may prefer squashing commits for which this does not hold.

All commits should be atomic. The documentation will have strange structure during the file moves, but hugo won't fail because of this.

The only thing that I can think of that would break is the old `utils/build_docs.py` after `util/difgen` is edited and moved into `mdbook/difgen` as part of the `mdbook_doxygen.py` preprocessor commit. I think it is alright to break the old docs, once the new ones are buildable though.

**Comment by HU90m** [comment]
> Git still shows some files as deleted rather than moved. One example is `doc/rm/crossbar_tool/index.md`. If I understand correctly, the content of that file has been moved elsewhere. I don't think there's a problem with that, but with so many changes how can we be sure that no content was lost in this reorganization?

I have been running `git log --name-status` and checking deleted files that aren't `R100` or `A`. Any file which is `D` should be explained in the commit message.

The github UI is struggling with the combined diff because some moves happen in two steps, so that one file can be moved out of the way before another file is put in it's place. One should be able to check these cases with `git log --follow -- $file`. Your particular example is a very good one:

```git
commit 0080eb7816c5d28820d8ac18d4c5c2018933f594
Author: Hugo McNally <hugo.mcnally@gmail.com>
Date:   Wed Feb 8 12:19:53 2023 +0000

    [doc mv] `util/` doc files moved for new layout.
    
    No file has been modified, so that git can more easily track moved files.
    
    `util/_index.md` was deleted as `doc/rm/_index.md`
    was a better starting point for a `util` overview/summary.
    
    Signed-off-by: Hugo McNally <hugo.mcnally@gmail.com>
    Co-authored-by: Harry Callahan <hcallahan@lowrisc.org>
    Co-authored-by: James Wainwright <james.wainwright@lowrisc.org>

R100    doc/rm/_index.md        util/README.md
D       util/_index.md
R100    doc/rm/vendor_in_tool.md        util/doc/vendor.md
R100    util/dvsim/doc/testplanner.md   util/dvsim/README.md
R100    util/fpvgen/doc/index.md        util/fpvgen/README.md
R100    doc/rm/ipgen_tool.md    util/ipgen/README.md
R100    doc/rm/register_tool/index.md   util/reggen/README.md
R100    doc/rm/register_tool/reg_top.svg        util/reggen/doc/reg_top.svg
R100    util/reggen/doc/index.md        util/reggen/doc/setup_and_use.md
R100    doc/rm/register_tool/subreg_ext.svg     util/reggen/doc/subreg_ext.svg
R100    doc/rm/register_tool/subreg_rw.svg      util/reggen/doc/subreg_rw.svg
R100    doc/rm/register_tool/subreg_rw0c.svg    util/reggen/doc/subreg_rw0c.svg
R100    doc/rm/register_tool/subreg_rw1c.svg    util/reggen/doc/subreg_rw1c.svg
R100    doc/rm/register_tool/subreg_rw1s.svg    util/reggen/doc/subreg_rw1s.svg
R100    doc/rm/register_tool/subreg_shadow.svg  util/reggen/doc/subreg_shadow.svg
R100    doc/rm/crossbar_tool/index.md   util/tlgen/README.md
R100    doc/rm/crossbar_tool/crossbar_example_1.svg     util/tlgen/doc/crossbar_example_1.svg
R100    doc/rm/crossbar_tool/crossbar_example_2.svg     util/tlgen/doc/crossbar_example_2.svg
R100    doc/rm/crossbar_tool/crossbar_example_3.svg     util/tlgen/doc/crossbar_example_3.svg
R100    doc/rm/crossbar_tool/crossbar_example_4.svg     util/tlgen/doc/crossbar_example_4.svg
R100    doc/rm/crossbar_tool/crossbar_example_5.svg     util/tlgen/doc/crossbar_example_5.svg
R100    doc/rm/topgen_tool.md   util/topgen/README.md
R100    util/uvmdvgen/doc/index.md      util/uvmdvgen/README.md

commit b7213f58b07f22949a3dec84bc0d41577e3410d0
Author: Hugo McNally <hugo.mcnally@gmail.com>
Date:   Wed Feb 8 12:10:07 2023 +0000

    [doc rm] removed file before moves for new docs
    
    `util/tlgen/README.md` contains no useful contents and will be replaced
    with the content of `doc/rm/crossbar_tool/index.md`.
    
    To make tracking moved files easier.
    This is being deleted in a separate commit.
    
    Signed-off-by: Hugo McNally <hugo.mcnally@gmail.com>

D       util/tlgen/README.md
```

File: util/autogen_md.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        match_end = END_MARKER_PATTERN.search(content, pos)
+        if match_end is None:
+            sys.exit(f"Error in {filepath}: start marker with command '{cmd}' without end marker")
+
+        print(f"In {filepath}: running '{cmd}'")
+        if dry_run:
+            # don't run
+            pos = match_end.end(0)
+            continue
+        res = subprocess.run(cmd, shell=True, capture_output=True)
```
Comment (timothytrippel) [inline util/autogen_md.py:53]: We will want to be careful where/how we execute this script, since we are potentially passing unsanitized inputs to the shell. Specifically, if someone submits a PR with an AUTOGEN block in it, any command could get executed (i.e., shell injection). If this only runs post-merge, then we could do some sanitization steps pre-merge or have some CI checks/bots that flag these for specific reviewer approval. Just wanted to point this out.

**Review by timothytrippel — APPROVED** [review]
This is large improvement over our current website. Big thanks to all that helped, well done!

(Since this isn't live yet, feel free to address any of my comments in a follow-up PR.)

File: util/mdbook/wavejson/default.js [hunk 23]
```diff
@@ -0,0 +1,3 @@
...,{id:'vmv-8-9'},['path',{d:'M9,0 20,0 20,20 9,20 6,10 z',class:'s14'}],['path',{d:'M3,0 0,0 0,20 3,20 6,10 z',class:'s13'}],['path',{d:'m0,0 3,0 6,20 11,0',class:'s1'}],['path',{d:'M0,20 3,20 9,0 20,0',class:'s1'}]],['g',{id:'vmv-9-9'},['path',{d:'M9,0 20,0 20,20 9,20 6,10 z',class:'s14'}],['path',{d:'M3,0 0,0 0,20 3,20 6,10 z',class:'s14'}],['path',{d:'m0,0 3,0 6,20 11,0',class:'s1'}],['path',{d:'M0,20 3,20 9,0 20,0',class:'s1'}]],['g',{id:'arrow0'},['path',{d:'m-12,-3 9,3 -9,3 c 1,-2 1,-4 0,-6 z',class:'s15'}],['path',{d:'M0,0 -15,0',class:'s16'}]],['marker',{id:'arrowhead',style:'fill:#0041c4',markerHeight:7,markerWidth:10,markerUnits:'strokeWidth',viewBox:'0 -4 11 8',refX:15,refY:0,orient:'auto'},['path',{d:'M0 -4 11 0 0 4z'}]],['marker',{id:'arrowtail',style:'fill:#0041c4',markerHeight:7,markerWidth:10,markerUnits:'strokeWidth',viewBox:'-11 -4 11 8',refX:-15,refY:0,orient:'auto'},['path',{d:'M0 -4 -11 0 0 4z'}]],['marker',{id:'tee',style:'fill:#0041c4',markerHeight:6,markerWidth:1,markerUnits:'strokeWidth',viewBox:'0 0 1 6',refX:0,refY:3,orient:'auto'},['path',{d:'M 0 0 L 0 6',style:'stroke:#0041c4;stroke-width:2'}]]],['g',{id:'waves'},['g',{id:'lanes'}],['g',{id:'groups'}]]];
```
Comment (timothytrippel) [inline util/mdbook/wavejson/default.js:1]: Another question: is there a license that needs to be re-iterated for this code? It looks like it is taken from another source.

File: doc/_index.md [hunk 24]
```diff
@@ -1,5 +0,0 @@
----
```
[the remark this answers is not in the document]
Comment (HU90m) [reply doc/_index.md:1]: I considered this but I found, when working on this, that keeping these apart makes it easier to track moved filed

File: util/tlgen/README.md [hunk 25]
```diff
@@ -1,4 +0,0 @@
-# TL-UL Crossbar RTL generator
```
[the remark this answers is not in the document]
Comment (HU90m) [reply util/tlgen/README.md:1]: Similar thoughts as https://github.com/lowRISC/opentitan/pull/17323#discussion_r1121753996

File: doc/introduction.md (hunk 5 above)
Comment (HU90m) [reply doc/introduction.md:36]: https://github.com/lowRISC/opentitan/pull/17323#discussion_r1121768123

File: doc/project_governance/development_stages.md (hunk 8 above)
Comment (HU90m) [reply doc/project_governance/development_stages.md:2]: Nope, just a hugo-ism

File: hw/formal/README.md (hunk 10 above)
Comment (HU90m) [reply hw/formal/README.md:3]: Yep, only files in the SUMMAR.md are in the table of contents. (Still technically accessible until mdbook ignore is sorted.)

File: hw/formal/README.md (hunk 11 above)
Comment (HU90m) [reply hw/formal/README.md:3]: Could we fix this in a future PR with other content changes?

File: util/mdbook-doxygen (hunk 13 above)
Comment (HU90m) [reply util/mdbook-doxygen]: Done

File: SUMMARY.md (hunk 16 above)
Comment (HU90m) [reply SUMMARY.md:16]: https://github.com/lowRISC/opentitan/pull/17323#discussion_r1121777941

File: book.toml (hunk 20 above)
Comment (HU90m) [reply book.toml:30]: Done

File: util/site/build-docs.sh (hunk 1 above)
Comment (HU90m) [reply util/site/build-docs.sh]: They are now documented but not validated

File: util/site/build-docs.sh (hunk 2 above)
Comment (HU90m) [reply util/site/build-docs.sh]: Yeah, I think `export`s are easier to read. I like `env` because it keeps environment variables scoped to the command they affect. So, I'm ok with both co-existing, but don't have particularly strong opinions either way.

File: util/site/build-docs.sh (hunk 3 above)
Comment (HU90m) [reply util/site/build-docs.sh:131]: Yes, but the `${proj_root}`  allows the script to work when run from anywhere. The `command` argument in `book.toml` is not relative to the book root :disappointed:

File: util/mdbook-reggen (hunk 19 above)
Comment (HU90m) [reply util/mdbook-reggen]: This is no longer appears in the history

File: util/mdbook/utils.py [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+def change_link_ext(
+        file_list: Set[Path],
+        content: str,
+        new_suffix: str,
+        book_root: Path,
+        page_path: Path,
+) -> str:
+    def suffix_swap(match: re.Match) -> str:
+        """Swaps the extension of the file being linked to if it is a ip block config."""
+        # print(match.group(0), file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (HU90m) [reply util/mdbook/utils.py:25]: Good spot! It's now gone.

File: util/mdbook/wavejson/default.js (hunk 23 above)
Comment (HU90m) [reply util/mdbook/wavejson/default.js:5]: Yes! Thank you, added a license headers to code taken from https://github.com/wavedrom/wavedrom

File: util/autogen_md.py (hunk 22 above)
Comment (HU90m) [reply util/autogen_md.py:53]: Yes, good to point out. It would be good to make it clear to reviewers that these commands are executed by linters

**Comment by HU90m** [comment]
Just pushed some changes.

TLDR: Ready to merge once render documentation CI check is fixed (and it is approved).

* Addressed most reviews.
* Cherry picked commits from https://github.com/lowRISC/opentitan/pull/17343
* Re-based onto latest master.
* Reworked the history into fewer commits.
* Fixed lint errors
  * flake8
  * trailing whitespaces
  * license checks
    * **NB:** Added MIT license to wavedrom files in `utils/mdbook/wavejson`.
  * ascii chars
    * **NB:** I have added `min.js` to the ignored files for ascii check. Reasoning above ignore.
  * Render documentation is the only lint that fails.
    * @hcallahan-lowrisc is this expected? How do we fix it?

**Comment by hcallahan-lowrisc** [comment]
> Just pushed some changes.
> 
> TLDR: Ready to merge once render documentation CI check is fixed (and it is approved).


@HU90m Awesome! Nice work. Thanks for getting all the fixes in order.

>   * Render documentation is the only lint that fails.
>     
>     * @hcallahan-lowrisc is this expected? How do we fix it?

That lint (defined here : https://github.com/lowRISC/opentitan/blob/master/azure-pipelines.yml#L130) just tries to build the docs the old way, so kinda expected that it fails.
We could address in a follow-up issue, or remove the lint temporarily, or actually fix the CI job to build the new docs site, but that one might be a little harder as I'm not sure if the Azure environment has the necessary deps for the new site.

**Review by moidx — APPROVED** [review]
LGTM! It is great to see regular markdown in the repo! I also like having `SUMMARY.md` as a way to organize the content hierarchy. The new style looks amazing and I find the new dashboards and easier to present to different types of audiences. 

Thanks for leading this big and impactful effort. This is already bringing the documentation on a whole new level! I look forward to seeing where you will take this in future.

**Review by jwnrt — APPROVED** [review]

**Review by hcallahan-lowrisc — APPROVED** [review]

File: doc/contributing/hw/methodology.md (hunk 4 above)
Comment (marnovandermaas) [reply doc/contributing/hw/methodology.md:227]: Splitting user guides into different books is essentially a content change though. This shouldn't be blocking to merging this PR, but we shouldn't forget to fix this in the future.

**Review by marnovandermaas — APPROVED** [review]
Thanks for the work and addressing most of my comments.

**Comment by GregAC** [comment]
Bypassing CI as the failure is unrelated to this change, great work everyone!
